Always give the pod owner control access

joachimvh opened the pull request:
Closes #720 

Depends on #936 

Some issues with this solution, some of them caused by architecture that might have to change:
1. Several components need things from the AccountStore but use a different key to access it. The login handler uses the email address while this new authorizer uses the WebID. This has been solved by allowing settings to be accessed through the WebID but an ideal solution would be to have a storage that can be queried (a triple store for example). But adding that while using the ResourceStore as backend and keeping it efficient is not going to be easy/possible.
2. The WAC-Allow headers are not going to be correct (or present event) for request that get handled by this authorizer. That is because a) this authorizer doesn't check the other permissions and b) this authorizer is WebACL-independent. Which means it returns an Authorization object with an empty `addMetadata` function. We will probably need an extra abstraction layer here somewhere in the future, several of our authorization components are too tightly coupled with the WebACL behaviour. Our internal permissions are probably also too tightly coupled to acl permissions.
3. Due to the account changes it is now always required to provide a password when creating a pod, even if you're not registering with the IDP. There are ways around it, but at some point in the future we probably want to provide some extra pod options which could be accessed using the same login details.
4. Also due to the account changes, it is no longer to have multiple accounts linked to the same WebID. There are ways around it if this is really an issue, but I think this is acceptable.
5. Since the new authorizer needs the AccountStore, it is now impossible to have a config that used the default authorizer without also importing the IDP handler config (see the change to the LDP integration config). I could potentialy add a new authorization config option that excludes this new authorizer for setups that do not want to import the IDP component.

RubenVerborgh commented on src/authorization/OwnershipAuthorizer.ts:
```diff
@@ -0,0 +1,41 @@
+    if (!credentials.webId) {
+      throw new NotImplementedHttpError('Only authenticated requests are supported.');
+    }
+    let settings: AccountSettings;
+    try {
+      settings = await this.accountStore.getSettings(credentials.webId);
+    } catch {
+      throw new NotImplementedHttpError('Only requests by registered WebIDs are supported.');
+    }
+    if (!settings.podBaseUrl || !identifier.path.startsWith(settings.podBaseUrl)) {
+      throw new NotImplementedHttpError('Only requests targeting the pod registered to this WebID are supported.');
+    }
+  }
+
+  public async handle(): Promise<Authorization> {
+    // If all checks in the canHandle function pass permission is always granted
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    return { addMetadata(): void {} };
+  }
```
I'm a bit surprised by the `canHandle`/`handle` division: I would expect `handle` to check whether the requesting agent is the owner of this pod. (So probably a reverse check.)

joachimvh replied:
The reason for this is because of some of the issues that you raise. It is written as it is so in the case this check fails (it's not the owner, the request is not for control permissions, etc.) the next handler will be called (the main acl authorizer).

RubenVerborgh replied:
Agreed.

RubenVerborgh replied:
To be precise: I agree with the code as-is, namely that it only has something to say in the specific case where an owner asks for `Control` access.

RubenVerborgh commented on src/authorization/OwnershipAuthorizer.ts:
```diff
@@ -0,0 +1,41 @@
+
+  public async handle(): Promise<Authorization> {
+    // If all checks in the canHandle function pass permission is always granted
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    return { addMetadata(): void {} };
```
```suggestion
    return { addMetadata(): void { /* no metadata */ } };
```

RubenVerborgh reviewed (changes requested):
2 and 4 are blockers for me.

But I think there's an even more fundamental problem: the ownership handler should be _combined_ with the WebACL authorizer. So the ownership handler should **add** `Control` permissions for the pod owner, rather than _only_ giving those permissions.

Right now, as an owner, I will never get read or write permissions, if I interpret this correctly.

So we either need ownership to be a wrapper, or all authorizers should take an object which they augment. (For instance, ownership could remove the need for "control" to be checked down the line, if it has already checked it.)

RubenVerborgh commented:
So we'll probably also need an integration test for this case.

RubenVerborgh commented:
This _might_ actually make the case for refactoring auth from pass/fail to returning the allowed modes instead. Then we could build a wrapping CombinedAuthenticationHandler (instead of waterfall) that adds allowed permissions together.

Alternatively, can we implementat owner access as rule, just like we've implemented groups?

joachimvh commented on test/integration/Identity.test.ts:
```diff
@@ -326,6 +334,43 @@ describe('A Solid server with IDP', (): void => {
       expect(res.status).toBe(205);
     });
+
+    it('always has control over data in the pod.', async(): Promise<void> => {
```
This is the new integration test

joachimvh commented:
> 
> 
> 2 and 4 are blockers for me.

To clarify 2: the headers are only going to be incorrect in the case the authorizer gets used, so specifically a control request done by the owner, all other requests are still going to have correct headers. But not perfect indeed.

4 can probably be resolved by storing an array of settings per WebID instead of a single object as is being done now, and then change the account store interface to return an array as wel instead. Would have to think about how to update preferences then but should be possible, just a bit more hassle perhaps.

> 
> But I think there's an even more fundamental problem: the ownership handler should be _combined_ with the WebACL authorizer. So the ownership handler should **add** `Control` permissions for the pod owner, rather than _only_ giving those permissions.
> 
> Right now, as an owner, I will never get read or write permissions, if I interpret this correctly.

You will because the `canHandle` call of this new authorizer is going to fail (since the permissions object contains non-control permissions), so the main acl authorizer will catch the request. The only issue would be if you do a request that requires both control and read permissions to your own pod, and the acl file only specifies that you have read permissions. But in practice such a request is (currently) impossible since it's always either one or the other, although that might be different for other authorizers.

> 
> So we either need ownership to be a wrapper, or all authorizers should take an object which they augment. (For instance, ownership could remove the need for "control" to be checked down the line, if it has already checked it.)

There might be something there.


> 
> 
> So we'll probably also need an integration test for this case.
> Perhaps see if owners can remove their own read/write (control??) access, and then restore it.

There is one, I added a comment to indicate where.

> 
> 
> This _might_ actually make the case for refactoring auth from pass/fail to returning the allowed modes instead. Then we could build a wrapping CombinedAuthenticationHandler (instead of waterfall) that adds allowed permissions together.

We could do that. One advantage now is that no acl check happens in the case an owner requires control permissions, which removes that backend call, but that's a very specific case so not that relevant. One issue is that for the WAC-Allow headers we need to know both the permissions that are allowed for the current request AND also the permissions required for a public user. We can just return both of course, but that might make our implementation a bit more acl-specific again.

> Alternatively, can we implement owner access as rule, just like we've implemented groups? (but yeah, it's not WebACL-specific)

Yea not a fan of this because of reasons mentioned above.

RubenVerborgh commented:
>  The only issue would be if you do a request that requires both control and read permissions to your own pod, and the acl file only specifies that you have read permissions. But in practice such a request is (currently) impossible

Right, I had not computed that indeed. They are currently mutually exclusive, my bad.

> > This _might_ actually make the case for refactoring auth from pass/fail to returning the allowed modes instead. Then we could build a wrapping CombinedAuthenticationHandler (instead of waterfall) that adds allowed permissions together.
> 
> We could do that.

I think actually this is my preference; because it also makes `WAC-Allow` independent of any specific authorizers.
The other consideration here is https://github.com/solid/web-access-control-spec/issues/97#issuecomment-913227652

Would you be okay with creating a new issue to refactor authorization to return allowed modes?
And then we rebase this one off it?

joachimvh commented:
I refactored this PR to be in line with the changes from #953. Going to look into points 2 and 4 mentioned above.

Point 2 (having correct WAC-Allow headers) might already be fixed due to this refactor. The headers on non-acl resources are now going to completely match what is defined in relevant acl resource. The only thing missing is that owners will not see the `control` value in the header if it's not specified in the acl resource, even though they do have those permissions. 

This can be fixed, but that would mean that we would have to do a call to the account store for every incoming request. I specifically made it so this currently only happens for requests targeting acl resources since this is another call to the backend resource store.

RubenVerborgh commented:
> The only thing missing is that owners will not see the `control` value in the header if it's not specified in the acl resource, even though they do have those permissions.
> 
> This can be fixed, but that would mean that we would have to do a call to the account store for every incoming request. I specifically made it so this currently only happens for requests targeting acl resources since this is another call to the backend resource store.

Sounds good to me. Will review current state now.

RubenVerborgh commented on src/authorization/AclOwnershipReader.ts:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Allows full access on ACL resources if the request is being made by the owner of the pod containing that resource.
+ */
+export class AclOwnershipReader extends PermissionReader {
```
Confusing name though.

`OwnerPermissionReader` or `PodOwnerPermissionReader`?

joachimvh replied:
I wanted to have `acl` in the name since this class is only going to work with acl authentication because of the `control` access mode.

RubenVerborgh replied:
Yeah but ACL points to the whole WebACL system, not specifically to the WebACL document (which is what you want here).

`ControlPermissionReader`?

But still think the "owner" labels are more descriptive.

joachimvh replied:
You could call it premature (name) optimization, but I was just thinking of the future when we also want to give the owner access when using a different auth system than acl. We would need a reader with a different name there.

RubenVerborgh replied:
Then I'd personally go for `OwnerPermissionReader`, but your call!

RubenVerborgh commented on src/authorization/AclOwnershipReader.ts:
```diff
@@ -0,0 +1,52 @@
+import { PermissionReader } from './PermissionReader';
+
+/**
+ * Allows full access on ACL resources if the request is being made by the owner of the pod containing that resource.
```
```suggestion
 * Allows control access if the request is being made by the owner of the pod containing the resource.
```

RubenVerborgh commented on src/authorization/AclOwnershipReader.ts:
```diff
@@ -0,0 +1,52 @@
+      return {};
+    }
+    if (!credentials.agent?.webId) {
+      // Only authenticated agents are supported
```
The `supported` language might create `canHandle` confusion.

RubenVerborgh commented on src/authorization/AclOwnershipReader.ts:
```diff
@@ -0,0 +1,52 @@
+  private readonly accountStore: AccountStore;
+  private readonly aclStrategy: AuxiliaryIdentifierStrategy;
+
+  public constructor(accountStore: AccountStore, aclStrategy: AuxiliaryIdentifierStrategy) {
```
Maybe logging would be interesting; could get confusing otherwise.

RubenVerborgh replied:
Implementation suggestion (if you think it makes sense): there are currently multiple paths returning `{}`. Perhaps these contents could instead be in a helper function that throws the exact reason for not granting the permissions; then the main `handle` catches the error, logs it as a `debug`, and returns `{}`. Then you could eliminate the comments from the function as well, as they would essentially be the error messages. (But that also holds if you just log them.)

joachimvh replied:
It's something I also wasn't quite sure of, both ways have some merit. The reason why it was done like this is that all other PermissionReaders can handle all input. Also because this one can handle the input, it just provides no permissions. But it's definitely not clearcut here imo.

RubenVerborgh replied:
> The reason why it was done like this is that all other PermissionReaders can handle all input.

No no, definitely keep that!

What I meant is:

```javascript
function canHandle() {
  try {
    this.verifyPodOwner();
  }
  catch (error) {
    this.logger.debug(`No pod owner Control permissions: ${error.message}`);
    return {};
  }
  return controlPermissions;
}
```

where `verifyPodOwner` contains the logic currently in `handle`, but then throws an explicit reason instead of returning `{}`:

```typescript
  public async handle({ credentials, identifier }: PermissionReaderInput): Promise<PermissionSet> {
    if (!this.aclStrategy.isAuxiliaryIdentifier(identifier)) {
      throw new Error("Not an ACL resource");
    }
```

So `handle` itself never throws.

joachimvh replied:
Ah sure, that works (I'm assuming you meant to write `function handle(input)` above).

RubenVerborgh replied:
Just a suggestion though. (Indeed!)

RubenVerborgh replied:
Just a suggestion though. (Indeed!)

RubenVerborgh commented on src/identity/interaction/email-password/handler/LoginHandler.ts:
```diff
@@ -23,6 +24,10 @@ export class LoginHandler extends InteractionHandler {
     const webId = await this.accountStore.authenticate(email, password);
+    const settings = await this.accountStore.getSettings(webId);
+    if (!settings.useIdp) {
+      throw new BadRequestHttpError('This account is not registered for identification');
```
I find the error message hard to understand, but couldn't make a suggestion as I don't know what it means exactly. Something like "This server is not an identity provider for the account xxx"?

joachimvh replied:
That is what it means yes, there is an account but it is only registered for pod management, not for IDP usage.

RubenVerborgh replied:
Okay, let's update the error message and also add a comment at that location in the code, explaining the above.

RubenVerborgh commented on src/identity/interaction/email-password/util/RegistrationManager.ts:
```diff
@@ -123,8 +123,13 @@ export class RegistrationManager {
     const trimmedEmail = this.trimString(email);
     assert(trimmedEmail && emailRegex.test(trimmedEmail), 'Please enter a valid e-mail address.');
 
+    const trimmedPassword = this.trimString(password);
```
Do we really trim passwords? Is that a documented practice?

joachimvh replied:
It's not documented, I added it (and pod name trimming) at some point in the past to be consistent with how we already trim WebIDs.

RubenVerborgh replied:
I think all the other trimming is fine; it's just that we don't say that spaces will be trimmed from passwords. This might wreak havoc on password managers, password length and complexity checkers, etc. So would _not_ trim password, but would trim all others.

RubenVerborgh commented on templates/identity/email-password/register-partial.html.ejs:
```diff
@@ -107,13 +107,6 @@
       </li>
     </ol>
   </div>
-  <div id="noPasswordForm" class="hidden">
```
Could you just explain this one? Forgot why it's no longer needed.

joachimvh replied:
Point 3 in the first comment way above:
> Due to the account changes it is now always required to provide a password when creating a pod, even if you're not registering with the IDP. There are ways around it, but at some point in the future we probably want to provide some extra pod options which could be accessed using the same login details.

RubenVerborgh commented on test/integration/Identity.test.ts:
```diff
@@ -320,6 +328,43 @@ describe('A Solid server with IDP', (): void => {
+      });
+      expect(res.status).toBe(205);
+
+      // Owner locked is locked out
```
?

RubenVerborgh replied:
As in, can't parse the sentence.

But also, is that correct behavior? Can we point to the spec requirement?

joachimvh replied:
Well that's just a bad sentence, not sure what happened there. W

hat happens here is that the acl file has been updated so that they don't have access to the resource anymore (all modes have been removed), which is why they get a 403 when accessing the resource. So the spec requirement is the WebACL spec. Due to the new permission reader they can update the acl resource though.

RubenVerborgh reviewed (approved):
Great work!

RubenVerborgh commented on src/authorization/AclOwnershipReader.ts:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  public async handle({ credentials, identifier }: PermissionReaderInput): Promise<PermissionSet> {
+    if (!this.aclStrategy.isAuxiliaryIdentifier(identifier)) {
```
PS I assume this case is for performance reasons?  (If so, let's comment that.)
Because the `Control` access would still be given to non-ACL resources.

joachimvh replied:
Yes, otherwise there would be a call to the storage for every request, now it's just for every request targeting an acl resource.

RubenVerborgh replied:
Check, let's just explain that in comment then.

RubenVerborgh commented on src/authorization/AclOwnershipReader.ts:
```diff
@@ -0,0 +1,52 @@
+    }
+    let settings: AccountSettings;
+    try {
+      settings = await this.accountStore.getSettings(credentials.agent.webId);
```
If too expensive (re: performance above), maybe we can consider a cache.

joachimvh replied:
It might be interesting to have some sort of cache for the key/value storages. Did we already have a discussion about this? Because I feel like we did. We would need events to update the caches then. Might even allow us to use an N3 store for the data and have events contain info which data to add/remove.

RubenVerborgh replied:
Yeah we did indeed. But not a necessity then; we know we can.
